internal/nix/search.go

```diff
@@ -106,3 +111,88 @@ func searchSystem(url, system string) (map[string]*Info, error) {
 	}
 	return parseSearchResults(out), nil
 }
+
+type CachedSearchResult struct {
+	Results map[string]*Info `json:"results"`
+	// Query is added easier to grep for debuggability
+	Query string `json:"query"`
+}
+
+var allowableQuery = regexp.MustCompile("^github:NixOS/nixpkgs/[0-9a-f]{40}#[^#]+$")
+
+func isAllowableQuery(query string) bool {
+	return allowableQuery.MatchString(query)
+}
+
+// SearchNixpkgsAttribute is a wrapper around searchSystem that caches results.
+// NOTE: we should be very conservative in where we use this function. `nix search`
+// accepts generalized `installable regex` as arguments but is slow. For certain
+// queries of the form `nixpkgs/<commit-hash>#attribute`, we can know for sure that
+// once `nix search` returns a valid result, it will always be the very same result.
+// Hence we can cache it locally and answer future queries fast, by not calling `nix search`.
+func SearchNixpkgsAttribute(query string) (map[string]*Info, error) {
+	if !isAllowableQuery(query) {
+		return nil, errors.Errorf("invalid query: %s, must match regex: %s", query, allowableQuery)
+	}
+
+	key := cacheKey(query)
+
+	// Check if the query was already cached, and return the result if so
+	cache := filecache.New("devbox/nix", filecache.WithCacheDir(xdg.CacheSubpath("")))
+	if cachedResults, err := cache.Get(key); err == nil {
+		var results map[string]*Info
+		if err := json.Unmarshal(cachedResults, &results); err != nil {
+			return nil, err
+		}
+		return results, nil
+	} else if !filecacheNeedsUpdate(err) {
+		return nil, err // genuine error
+	}
+
+	// If not cached, or an update is needed, then call searchSystem
+	infos, err := searchSystem(query, "" /*system*/)
+	if err != nil {
+		return nil, err
+	}
+
+	// Save the results to the cache
+	marshalled, err := json.Marshal(infos)
+	if err != nil {
+		return nil, err
+	}
+	// TODO savil: add a SetForever API that does not expire. Time based expiration is not needed here
+	// because we're caching results that are guaranteed to be stable.
+	// TODO savil: Make filecache.cache a public struct so it can be passed into other functions
+	const oneYear = 12 * 30 * 24 * time.Hour
+	if err := cache.Set(key, marshalled, oneYear); err != nil {
+		return nil, err
+	}
+
+	return infos, nil
+}
+
+// read as: filecache.NeedsUpdate(err)
+// TODO savil: this should be implemented in the filecache package
+func filecacheNeedsUpdate(err error) bool {
```

Comment:
fwiw I would just name this `isNotFoundOrExpired`

Maybe in filecache, using new `errors.Join` we can return an error that is both NotFound and also something like `CacheMiss`

so `return errors.Join(NotFound, CacheMiss)`

and then here we just check if it is `CacheMiss`

Reply:
yeah, that sgtm

Reply:
named it as: `filecacheIsCacheMiss`